File: frontend/src/types/course.ts
// Course Level Types
export type CourseLevelId =
  | 'awareness'
  | 'foundational'
  | 'basic'
  | 'intermediate'
  | 'advanced'
  | 'expert'
  | 'senior';

export interface CourseLevel {
  id: CourseLevelId;
  name: string;
  description: string;
  hoursRange: {
    min: number;
    max: number;
  };
  targetAudience: string;
}

// Course Thematic Types
export type CourseThematicId =
  | 'defence-security'
  | 'intelligence'
  | 'policing'
  | 'leadership'
  | 'crisis-response'
  | 'resilience'
  | 'personal-skills'
  | 'user-defined';

export interface CourseThematic {
  id: CourseThematicId;
  name: string;
  description: string;
  keywords: string[];
  color: string;
}

// Course Status Types
export type CourseStatusCode =
  | 'DRAFT'
  | 'IN_PROGRESS'
  | 'REVIEW'
  | 'APPROVED'
  | 'PUBLISHED'
  | 'ARCHIVED';

export interface CourseStatus {
  code: CourseStatusCode;
  label: string;
  color: string;
}

// Learning Objective Types
export interface LearningObjective {
  id: string;
  type: 'terminal' | 'enabling';
  text: string;
  parentId?: string;
  order: number;
}

// Module and Lesson Types
export interface Lesson {
  id: string;
  number: number;
  title: string;
  duration: number;
  content: string;
  keyPoints: string[];
  activities: string[];
}

export interface Module {
  id: string;
  number: number;
  title: string;
  description: string;
  duration: number;
  lessons: Lesson[];
}

// Assessment Types
export type AssessmentType =
  | 'Multiple Choice Quiz'
  | 'Written Assessment'
  | 'Practical Exercise'
  | 'Scenario-Based Assessment'
  | 'Performance Evaluation'
  | 'Portfolio Review'
  | 'Peer Assessment'
  | 'Self-Assessment'
  | 'Oral Examination'
  | 'Capstone Project';

export interface Assessment {
  id: string;
  type: AssessmentType;
  title: string;
  description: string;
  criteria: string[];
  passingScore: number;
  duration: number;
}

// Delivery Method Types
export type DeliveryMethod =
  | 'Instructor-Led Training (ILT)'
  | 'Virtual Instructor-Led Training (VILT)'
  | 'Self-Paced eLearning'
  | 'Blended Learning'
  | 'On-the-Job Training'
  | 'Workshop'
  | 'Seminar'
  | 'Simulation-Based Training'
  | 'Mobile Learning';

// Course Metadata
export interface CourseMetadata {
  createdDate: string;
  updatedDate: string;
  author: string;
  reviewer?: string;
  organization?: string;
  version: string;
}

// Main Course Interface
export interface Course {
  id: string;
  code: string;
  title: string;
  level: CourseLevelId;
  thematic: CourseThematicId;
  customThematic?: string;
  status: CourseStatusCode;
  description: string;
  overview: string;
  targetAudience: string;
  duration: number;
  deliveryMethod: DeliveryMethod;
  learningObjectives: LearningObjective[];
  modules: Module[];
  assessments: Assessment[];
  metadata: CourseMetadata;
}

// Course Creation/Update DTOs
export interface CreateCourseDTO {
  title: string;
  code: string;
  level: CourseLevelId;
  thematic: CourseThematicId;
  customThematic?: string;
  description: string;
  overview: string;
  targetAudience: string;
  duration: number;
  deliveryMethod: DeliveryMethod;
  author: string;
  organization?: string;
}

export interface UpdateCourseDTO extends Partial<CreateCourseDTO> {
  id: string;
}

// AI Generation Types
export interface AIGenerationRequest {
  courseId: string;
  generationType: 'objectives' | 'modules' | 'assessments' | 'full';
  context?: string;
  preferences?: Record<string, unknown>;
}

export interface AIGenerationResponse {
  success: boolean;
  data?: Partial<Course>;
  message?: string;
  tokens_used?: number;
}

// Chat Message Types
export interface ChatMessage {
  id: string;
  role: 'user' | 'assistant' | 'system';
  content: string;
  timestamp: string;
}

// Export Types
export type ExportFormat = 'json' | 'pdf' | 'docx' | 'scorm';

export interface ExportRequest {
  courseId: string;
  format: ExportFormat;
  includeMetadata: boolean;
}

export interface ExportResponse {
  success: boolean;
  downloadUrl?: string;
  message?: string;
}

File: backend/app/models/course.py
"""Course data models."""

from pydantic import BaseModel, Field
from typing import Optional, List
from enum import Enum
from datetime import datetime


class CourseLevelEnum(str, Enum):
    """Course difficulty levels."""
    AWARENESS = "awareness"
    FOUNDATIONAL = "foundational"
    BASIC = "basic"
    INTERMEDIATE = "intermediate"
    ADVANCED = "advanced"
    EXPERT = "expert"
    SENIOR = "senior"


class CourseThematicEnum(str, Enum):
    """Course thematic areas."""
    DEFENCE_SECURITY = "defence-security"
    INTELLIGENCE = "intelligence"
    POLICING = "policing"
    LEADERSHIP = "leadership"
    CRISIS_RESPONSE = "crisis-response"
    RESILIENCE = "resilience"
    PERSONAL_SKILLS = "personal-skills"
    USER_DEFINED = "user-defined"


class CourseStatusEnum(str, Enum):
    """Course status codes."""
    DRAFT = "DRAFT"
    IN_PROGRESS = "IN_PROGRESS"
    REVIEW = "REVIEW"
    APPROVED = "APPROVED"
    PUBLISHED = "PUBLISHED"
    ARCHIVED = "ARCHIVED"


class DeliveryMethodEnum(str, Enum):
    """Course delivery methods."""
    ILT = "Instructor-Led Training (ILT)"
    VILT = "Virtual Instructor-Led Training (VILT)"
    SELF_PACED = "Self-Paced eLearning"
    BLENDED = "Blended Learning"
    OJT = "On-the-Job Training"
    WORKSHOP = "Workshop"
    SEMINAR = "Seminar"
    SIMULATION = "Simulation-Based Training"
    MOBILE = "Mobile Learning"


class LearningObjective(BaseModel):
    """Learning objective model."""
    id: str
    type: str = Field(..., pattern="^(terminal|enabling)$")
    text: str
    parent_id: Optional[str] = None
    order: int = 0


class Lesson(BaseModel):
    """Lesson model."""
    id: str
    number: int
    title: str
    duration: int = 0
    content: str = ""
    key_points: List[str] = []
    activities: List[str] = []


class Module(BaseModel):
    """Module model."""
    id: str
    number: int
    title: str
    description: str = ""
    duration: int = 0
    lessons: List[Lesson] = []


class Assessment(BaseModel):
    """Assessment model."""
    id: str
    type: str
    title: str
    description: str = ""
    criteria: List[str] = []
    passing_score: int = 70
    duration: int = 0


class CourseMetadata(BaseModel):
    """Course metadata model."""
    created_date: str = Field(default_factory=lambda: datetime.now().isoformat())
    updated_date: str = Field(default_factory=lambda: datetime.now().isoformat())
    author: str = ""
    reviewer: Optional[str] = None
    organization: Optional[str] = None
    version: str = "1.0.0"


class CourseBase(BaseModel):
    """Base course model."""
    title: str
    code: str
    level: Optional[CourseLevelEnum] = None
    thematic: Optional[CourseThematicEnum] = None
    custom_thematic: Optional[str] = None
    description: str = ""
    overview: str = ""
    target_audience: str = ""
    duration: int = 0
    delivery_method: Optional[DeliveryMethodEnum] = None


class CourseCreate(CourseBase):
    """Course creation model."""
    author: str = ""
    organization: Optional[str] = None


class CourseUpdate(BaseModel):
    """Course update model."""
    title: Optional[str] = None
    code: Optional[str] = None
    level: Optional[CourseLevelEnum] = None
    thematic: Optional[CourseThematicEnum] = None
    custom_thematic: Optional[str] = None
    status: Optional[CourseStatusEnum] = None
    description: Optional[str] = None
    overview: Optional[str] = None
    target_audience: Optional[str] = None
    duration: Optional[int] = None
    delivery_method: Optional[DeliveryMethodEnum] = None
    learning_objectives: Optional[List[LearningObjective]] = None
    modules: Optional[List[Module]] = None
    assessments: Optional[List[Assessment]] = None


class Course(CourseBase):
    """Full course model."""
    id: str
    status: CourseStatusEnum = CourseStatusEnum.DRAFT
    learning_objectives: List[LearningObjective] = []
    modules: List[Module] = []
    assessments: List[Assessment] = []
    metadata: CourseMetadata = Field(default_factory=CourseMetadata)

    class Config:
        from_attributes = True
